File: platform/cli/src/commands/land.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as git from '../git';
import {
  loadConfig,
  getCurrentFeature,
  loadDiffsForFeature,
  saveDiff,
  saveFeature,
  saveUndoEntry,
  type DiffData,
} from '../workspace';
import {
  errorMessage,
  diffLabel,
  diffStatusIcon,
  nextStepHint,
} from '../format';

function parseDiffPosition(label: string): number | null {
  const match = label.match(/^[Dd](\d+)$/);
  if (!match) return null;
  return parseInt(match[1]!, 10);
}

export function registerLandCommand(program: Command): void {
  program
    .command('land')
    .description('Land the lowest approved diff onto the main codebase')
    .argument('[label]', 'Diff label (e.g., D1). Defaults to lowest un-landed.')
    .option('--force', 'Skip approval/CI checks')
    .action(async (label: string | undefined, opts: { force?: boolean }) => {
      try {
        const feature = getCurrentFeature();
        if (!feature) {
          console.log(chalk.red('No active feature.'));
          console.log(
            chalk.dim('Start one with: devpod feature "name"'),
          );
          process.exit(1);
          return;
        }

        const diffs = loadDiffsForFeature(feature).sort(
          (a, b) => a.position - b.position,
        );

        if (diffs.length === 0) {
          console.log(chalk.dim('No diffs to land.'));
          return;
        }

        // Find the target diff
        let targetDiff: DiffData;
        const unlandedDiffs = diffs.filter((d) => d.status !== 'landed');

        if (label) {
          const position = parseDiffPosition(label);
          if (position === null) {
            console.log(
              chalk.red(`Invalid diff label: ${label}. Use format D1, D2, etc.`),
            );
            process.exit(1);
            return;
          }

          const found = diffs.find((d) => d.position === position);
          if (!found) {
            console.log(chalk.red(`${diffLabel(position)} not found.`));
            process.exit(1);
            return;
          }

          // Must be the bottom (lowest position) un-landed diff
          if (unlandedDiffs.length > 0 && found.position !== unlandedDiffs[0]!.position) {
            console.log(
              chalk.red(
                `${diffLabel(position)} is not the lowest diff. Land ${diffLabel(unlandedDiffs[0]!.position)} first.`,
              ),
            );
            process.exit(1);
            return;
          }

          targetDiff = found;
        } else {
          if (unlandedDiffs.length === 0) {
            console.log(chalk.dim('All diffs already landed.'));
            return;
          }
          targetDiff = unlandedDiffs[0]!;
        }

        // Check status
        if (!opts.force) {
          if (
            targetDiff.status !== 'approved' &&
            targetDiff.ci !== 'passed'
          ) {
            console.log(
              chalk.yellow(
                `${diffLabel(targetDiff.position)} has not been approved and CI has not passed.`,
              ),
            );
            console.log(
              chalk.dim('Use --force to land anyway.'),
            );
            process.exit(1);
            return;
          }
        }

        // Save undo entry
        saveUndoEntry({
          action: 'land',
          timestamp: new Date().toISOString(),
          refBefore: git.getHeadSha(),
          description: `Land ${diffLabel(targetDiff.position)}: ${targetDiff.title}`,
          data: {
            uuid: targetDiff.uuid,
            featureSlug: feature.slug,
          },
        });

        const config = loadConfig();
        const featureBranch = feature.branch;

        // Switch to main, pull latest
        git.switchBranch(config.defaultBranch);
        git.fetchMain();
        git.rebaseOnto(`origin/${config.defaultBranch}`);

        // Cherry-pick squash the diff's commit
        git.cherryPickSquash(targetDiff.commit);
        git.commit(targetDiff.title);

        // Push to main
        git.pushForce(config.defaultBranch);

        // Switch back to feature branch
        git.switchBranch(featureBranch);

        // Rebase remaining diffs onto main
        const remainingDiffs = unlandedDiffs.filter(
          (d) => d.position !== targetDiff.position,
        );
        if (remainingDiffs.length > 0) {
          git.rebaseOnto(`origin/${config.defaultBranch}`);

          // Update commit SHAs for remaining diffs
          const commits = git.getCommitsBetween(
            `origin/${config.defaultBranch}`,
            'HEAD',
          );
          const orderedCommits = [...commits].reverse();
          const sortedRemaining = remainingDiffs.sort(
            (a, b) => a.position - b.position,
          );
          for (
            let i = 0;
            i < sortedRemaining.length && i < orderedCommits.length;
            i++
          ) {
            sortedRemaining[i]!.commit = orderedCommits[i]!.sha;
            saveDiff(sortedRemaining[i]!);
          }
        }

        // Update landed diff
        targetDiff.status = 'landed';
        targetDiff.updated = new Date().toISOString();
        saveDiff(targetDiff);

        // Remove from feature's diffs list
        feature.diffs = feature.diffs.filter((id) => id !== targetDiff.uuid);
        if (feature.diffs.length === 0) {
          feature.status = 'complete';
        }
        saveFeature(feature);

        // Result
        console.log(
          `${chalk.green('\u2713')} Landed ${diffLabel(targetDiff.position)}: ${targetDiff.title}`,
        );

        if (feature.diffs.length === 0) {
          console.log(
            chalk.dim('  Feature complete! All diffs landed.'),
          );
          console.log(
            chalk.dim('  You can start a new feature with: devpod feature "name"'),
          );
        } else {
          const remainingLoaded = loadDiffsForFeature(feature);
          const stackParts = remainingLoaded.map((d) => {
            return `${diffLabel(d.position)} ${diffStatusIcon(d.status)}`;
          });
          console.log(
            chalk.dim(`  Remaining: ${stackParts.join(' \u2192 ')}`),
          );
        }

        console.log(nextStepHint('land'));
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}

File: platform/cli-go/internal/commands/land.go
package commands

import (
	"fmt"
	"strings"
	"time"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

func newLandCmd() *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:   "land [label]",
		Short: "Land the lowest approved diff onto the main codebase",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			feature := workspace.GetCurrentFeature()
			if feature == nil {
				fmt.Println(format.ErrorMsg("No active feature."))
				fmt.Println(format.DimText("Start one with: devpod feature \"name\""))
				return fmt.Errorf("no active feature")
			}

			diffs := sortDiffsByPosition(workspace.LoadDiffsForFeature(*feature))

			if len(diffs) == 0 {
				fmt.Println(format.DimText("No diffs to land."))
				return nil
			}

			// Find unlanded diffs
			var unlandedDiffs []workspace.DiffData
			for _, d := range diffs {
				if d.Status != "landed" {
					unlandedDiffs = append(unlandedDiffs, d)
				}
			}

			var targetDiff workspace.DiffData

			if len(args) > 0 {
				label := args[0]
				position, ok := parseDiffPosition(label)
				if !ok {
					msg := fmt.Sprintf("Invalid diff label: %s. Use format D1, D2, etc.", label)
					fmt.Println(format.ErrorMsg(msg))
					return fmt.Errorf("%s", msg)
				}

				var found *workspace.DiffData
				for i, d := range diffs {
					if d.Position == position {
						found = &diffs[i]
						break
					}
				}
				if found == nil {
					msg := fmt.Sprintf("%s not found.", format.DiffLabel(position))
					fmt.Println(format.ErrorMsg(msg))
					return fmt.Errorf("%s", msg)
				}

				// Must be the bottom (lowest position) un-landed diff
				if len(unlandedDiffs) > 0 && found.Position != unlandedDiffs[0].Position {
					msg := fmt.Sprintf("%s is not the lowest diff. Land %s first.",
						format.DiffLabel(position), format.DiffLabel(unlandedDiffs[0].Position))
					fmt.Println(format.ErrorMsg(msg))
					return fmt.Errorf("%s", msg)
				}

				targetDiff = *found
			} else {
				if len(unlandedDiffs) == 0 {
					fmt.Println(format.DimText("All diffs already landed."))
					return nil
				}
				targetDiff = unlandedDiffs[0]
			}

			// Check status
			if !force {
				if targetDiff.Status != "approved" && targetDiff.CI != "passed" {
					fmt.Println(format.WarnMsg(fmt.Sprintf(
						"%s has not been approved and CI has not passed.",
						format.DiffLabel(targetDiff.Position))))
					fmt.Println(format.DimText("Use --force to land anyway."))
					return fmt.Errorf("not approved")
				}
			}

			// Save undo entry
			headBefore, _ := git.GetHeadSHA()
			_ = workspace.SaveUndoEntry(workspace.UndoEntry{
				Action:      "land",
				Timestamp:   time.Now().UTC().Format(time.RFC3339),
				RefBefore:   headBefore,
				Description: fmt.Sprintf("Land %s: %s", format.DiffLabel(targetDiff.Position), targetDiff.Title),
				Data: map[string]interface{}{
					"uuid":        targetDiff.UUID,
					"featureSlug": feature.Slug,
				},
			})

			config := workspace.LoadConfig()
			featureBranch := feature.Branch

			// Switch to main, pull latest
			_ = git.SwitchBranch(config.DefaultBranch)
			_ = git.FetchMain()
			_ = git.RebaseOnto("origin/" + config.DefaultBranch)

			// Cherry-pick squash the diff's commit
			_ = git.CherryPickSquash(targetDiff.Commit)
			_, _ = git.Commit(targetDiff.Title)

			// Push to main
			_ = git.PushForce(config.DefaultBranch)

			// Switch back to feature branch
			_ = git.SwitchBranch(featureBranch)

			// Rebase remaining diffs onto main
			var remainingDiffs []workspace.DiffData
			for _, d := range unlandedDiffs {
				if d.Position != targetDiff.Position {
					remainingDiffs = append(remainingDiffs, d)
				}
			}

			if len(remainingDiffs) > 0 {
				_ = git.RebaseOnto("origin/" + config.DefaultBranch)

				// Update commit SHAs for remaining diffs
				commits, _ := git.GetCommitsBetween("origin/"+config.DefaultBranch, "HEAD")
				orderedCommits := make([]git.LogEntry, len(commits))
				for i, c := range commits {
					orderedCommits[len(commits)-1-i] = c
				}
				sortedRemaining := sortDiffsByPosition(remainingDiffs)
				for i := 0; i < len(sortedRemaining) && i < len(orderedCommits); i++ {
					sortedRemaining[i].Commit = orderedCommits[i].SHA
					_ = workspace.SaveDiff(sortedRemaining[i])
				}
			}

			// Update landed diff
			targetDiff.Status = "landed"
			targetDiff.Updated = time.Now().UTC().Format(time.RFC3339)
			_ = workspace.SaveDiff(targetDiff)

			// Remove from feature's diffs list
			var newDiffs []string
			for _, id := range feature.Diffs {
				if id != targetDiff.UUID {
					newDiffs = append(newDiffs, id)
				}
			}
			feature.Diffs = newDiffs
			if len(feature.Diffs) == 0 {
				feature.Status = "complete"
			}
			_ = workspace.SaveFeature(*feature)

			// Result
			fmt.Printf("%s Landed %s: %s\n", format.SuccessMsg("\u2713"), format.DiffLabel(targetDiff.Position), targetDiff.Title)
			fmt.Printf("  History: http://localhost:3000/diffs/%s\n", format.DimText(feature.Slug))

			if len(feature.Diffs) == 0 {
				fmt.Println(format.DimText("  Feature complete! All diffs landed."))
				fmt.Println(format.DimText("  You can start a new feature with: devpod feature \"name\""))
			} else {
				remainingLoaded := workspace.LoadDiffsForFeature(*feature)
				sorted := sortDiffsByPosition(remainingLoaded)
				var stackParts []string
				for _, d := range sorted {
					stackParts = append(stackParts, fmt.Sprintf("%s %s", format.DiffLabel(d.Position), format.DiffStatusIcon(string(d.Status))))
				}
				fmt.Println(format.DimText(fmt.Sprintf("  Remaining: %s", strings.Join(stackParts, " \u2192 "))))
			}

			fmt.Println(format.NextStepHint("land"))
			return nil
		},
	}

	cmd.Flags().BoolVar(&force, "force", false, "Skip approval/CI checks")

	return cmd
}
